proposals/0166-dynamic-stack-frames.md

```diff
@@ -0,0 +1,173 @@
+pointer address to access those parameters.
+
+### Changes in the verifier
+
+The additional constraints for the verifier only encompass new rules for the 
+introduction of the R11 register. It can only appear coupled with a `add64` 
+(opcode `0x07`) instruction. If that is not the case, the verifier must throw 
+a `VerifierError::InvalidSourceRegister` when it is the source register of an 
+instruction and `VerifierError::InvalidDestinationRegister` when it is the 
+destination register of an instruction. Additionally, it must not be the 
+target register for the `callx` (opcode 0x9D) instructions, otherwise a 
+`VerifierError::InvalidRegister` must be raised. 
+
+The verification rules for the R10 register must remain unchanged. It must not 
+be the target register for the `callx` (opcode 0x9D) instructions, otherwise a 
+`VerifierError::InvalidRegister` must be raised. It must also not be the 
+destination register of non-store instructions, otherwise a 
+`VerifierError::CannotWriteToR10` must be thrown.
+
+
+### Identification of programs
+
+As per the description in SIMD-0161, programs compiled with dynamic stack 
+frames must contain the XX flag on their ELF header `e_flags` field.
+
+## Impact
+
+We foresee a positive impact in smart contract development. Developers won’t 
+need to worry about exceeding the maximum frame space allowed for a function 
+and won’t face any case of stack access violation if their code follows 
+conventional Rust safety rules. Likewise, when we update the Rust version of 
+our platform tools, developers will not have the burden of modifying their 
+contract just because the newer version is using more stack than the previous 
+one, often reaching the 4096 bytes limit. Refer to issues 
+[#1186](https://github.com/anza-xyz/agave/issues/1186) and 
+[#1158](https://github.com/anza-xyz/agave/issues/1158).
+
+We also expect some improvements in program execution. For functions with no 
+stack usage, we will not emit the additional instruction that modifies R11, 
+saving some execution time. Furthermore, for function calls that handle more 
+than five arguments, there will be one less store and one less load operation 
+due to the new call convention.
+
+## Security Considerations
+
+Stack gaps will be disabled for dynamic stack frames to work. Stack gaps could 
+detect invalid accesses between two function frames, if the accessed address 
+would fall between them. With dynamic stack frames, all stack access will be 
+valid, provided that their address is within the allowed range. We already 
+allow functions to read and modify the memory inside the frame of other 
+functions, so removing the stack gaps should not bring any security 
+implications.
+
+Although one can change R11 to any value that fits in a 64-bit integer, every 
+memory access is verified, so there is no risk of invalid accesses from a 
+corrupt register.
+
+## Drawbacks
+
+Programs will consume more compute units, as most functions will include two 
```

Comment:
I'd slightly rephrase:

Negligible. Programs will consume a few more compute units, as most functions will include two 
 extra instructions: one to increment the stack pointer and another one to decrement it.

Reply:
Fixed.